src/base/datetime/calendar/index.ts

```diff
@@ -86,34 +87,101 @@ export class BaseDateTimeCalendar extends KucBase {
   }
 
   private _calculateBodyCalendarPosition() {
-    const baseDateTimeEl = this._baseCalendarGroupEl.parentElement
-      ?.parentElement;
-    if (!baseDateTimeEl) return;
+    const distance = this._calculateDistanceInput();
+    if (!distance) return;
+
+    const {
+      inputToBottom,
+      inputToTop,
+      inputToRight,
+      inputToLeft,
+      calendarHeight
+    } = distance;
+
+    if (inputToBottom < calendarHeight) {
+      if (inputToTop < 0 || inputToBottom > inputToTop) {
+        this._calculateCalendarPosition(inputToRight, inputToLeft, "bottom");
+        return;
+      }
+      this._calculateCalendarPosition(inputToRight, inputToLeft, "top");
+      return;
+    }
+    this._calculateCalendarPosition(inputToRight, inputToLeft, "bottom");
+  }
 
-    const dateInputEl = baseDateTimeEl.querySelector(
-      ".kuc-base-date__input"
-    ) as HTMLInputElement;
-    if (!dateInputEl) return;
+  private _calculateDistanceInput() {
+    if (!this.parentElement) return false;
 
-    const dateInputTop = dateInputEl.offsetTop;
-    const dateTop = dateInputTop + dateInputEl.offsetHeight;
+    const inputToBottom =
+      window.innerHeight - this.parentElement.getBoundingClientRect().bottom;
+    const inputToTop = this.parentElement.getBoundingClientRect().top;
+    const inputToRight =
+      document.body.clientWidth - this.parentElement.offsetLeft;
+    const inputToLeft = this.parentElement.getBoundingClientRect().left;
+    const calendarHeight = this._baseCalendarGroupEl.getBoundingClientRect()
+      .height;
 
-    const baseDatetimeCalendarEl = this._baseCalendarGroupEl.parentElement;
-    if (!baseDatetimeCalendarEl) return;
-    baseDatetimeCalendarEl.style.top = dateTop + "px";
-    baseDatetimeCalendarEl.style.left = dateInputEl.offsetLeft + "px";
+    return {
+      inputToBottom,
+      inputToTop,
+      inputToRight,
+      inputToLeft,
+      calendarHeight
+    };
+  }
```

Comment:
can you return the same type ? at row 113 you return boolean, and at here you return object